src/uct/ib/base/ib_iface.c

```diff
@@ -1204,53 +1187,73 @@ uct_ib_iface_select_sl(const uct_ib_iface_config_t *ib_config,
     }
 
+    ucs_string_buffer_init(&sl_mask_str);
+
     if (ib_config->sl == UCS_ULUNITS_AUTO) {
-        /* Select the first available SL with AR support, since a specific
-         * SL was not requested by user */
+        /* Select the first available SL from mask, since a specific SL was not
+         * requested by user */
         sl_from_mask = ucs_ffs64_safe(ooo_sl_mask);
         if (sl_from_mask != 64) {
-            sl = sl_from_mask;
-        } else if (ar_enable == UCS_YES) {
+            /* ooo_sl_mask wasn't empty, SL was selected */
+            *sl_p = sl_from_mask;
+        } else if (ar_enable != UCS_TRY) {
+            /* ooo_sl_mask was empty, SL wasn't selected */
             goto err;
+        } else {
+            /* ooo_sl_mask was empty, SL wasn't selected, but user requested
+             * to try to find the SL with AR support, select the default one */
+            *sl_p = uct_ib_iface_config_select_sl(ib_config);
+        }
+    } else if (!(ooo_sl_mask & UCS_BIT(ib_config->sl))) {
+        if (ar_enable != UCS_TRY) {
+            /* A specific SL wasn't found in ooo_sl_mask */
+            goto err;
+        } else {
+            /* A specific SL wasn't found in ooo_sl_mask, but user requested
+             * to try to find the SL with AR support, select the default one */
+            *sl_p = uct_ib_iface_config_select_sl(ib_config);
+        }
+    } else {
+        /* SL with (AR_ENABLE=yes) or without (AR_ENABLE=no) AR was found */
+        *sl_p                 = (uint8_t)ib_config->sl;
+        if (ar_enable == UCS_TRY) {
+            sl_ar_support_str = "with";
         }
-    } else if (!(ooo_sl_mask & UCS_BIT(ib_config->sl)) &&
-               (ar_enable == UCS_YES)) {
-        goto err;
     }
 
-out:
-    *sl_p = sl;
-    ucs_debug("SL=%u was selected, SLs with AR support = { %s } on %s",
-              sl, uct_ib_iface_sl_mask_str(ooo_sl_mask, &sl_mask_str),
+    ucs_debug("SL=%u %s AR was selected, SLs %s AR support = { %s } on %s",
+              *sl_p, sl_ar_support_str, sl_ar_support_str,
+              ucs_mask_str(ooo_sl_mask, &sl_mask_str),
               dev_port_name);
 out_str_buf_clean:
     ucs_string_buffer_cleanup(&sl_mask_str);
     return status;
 
 err:
+    ucs_assert(ar_enable != UCS_TRY);
     ucs_config_sprintf_ulunits(sl_str, sizeof(sl_str), &ib_config->sl, NULL);
-    ucs_error("AR support is requested for SL=%s, but %s on %s,"
-              " SLs with AR support = { %s }", sl_str,
-              ooo_sl_mask_is_detected ? "could not select SL with AR support" :
-              "could not detect SLs with AR. Please, set it manually",
-              dev_port_name, uct_ib_iface_sl_mask_str(ooo_sl_mask,
-                                                      &sl_mask_str));
+    ucs_config_sprintf_ternary(ar_enable_str, sizeof(ar_enable_str), &ar_enable,
+                               NULL);
+    ucs_error("AR=%s was requested for SL=%s, but %s %s AR on %s,"
```

Comment:
can you pls post few example error messages?

Reply:
```
[     INFO ] < AR=n was requested for SL=auto, but could not detect AR mask for SLs. Please, set SL manually without AR on ib_dev:1, SLs with AR support = { <none> }, SLs without AR support = { <none> } >
[     INFO ] < AR=n was requested for SL=8, but could not detect AR mask for SLs. Please, set SL manually without AR on ib_dev:1, SLs with AR support = { <none> }, SLs without AR support = { <none> } >
[     INFO ] < AR=n was requested for SL=8, but could not select SL without AR on ib_dev:1, SLs with AR support = { 8 }, SLs without AR support = { <none> } >
[     INFO ] < AR=n was requested for SL=8, but could not select SL without AR on ib_dev:1, SLs with AR support = { 8 }, SLs without AR support = { <none> } >
[     INFO ] < AR=y was requested for SL=auto, but could not select SL with AR on ib_dev:1, SLs with AR support = { <none> }, SLs without AR support = { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14 } >
[     INFO ] < AR=y was requested for SL=auto, but could not detect AR mask for SLs. Please, set SL manually with AR on ib_dev:1, SLs with AR support = { <none> }, SLs without AR support = { <none> } >
[     INFO ] < AR=y was requested for SL=8, but could not select SL with AR on ib_dev:1, SLs with AR support = { <none> }, SLs without AR support = { 8 } >
[     INFO ] < AR=y was requested for SL=8, but could not detect AR mask for SLs. Please, set SL manually with AR on ib_dev:1, SLs with AR support = { <none> }, SLs without AR support = { <none> } >
[     INFO ] < AR=y was requested for SL=8, but could not select SL with AR on ib_dev:1, SLs with AR support = { <none> }, SLs without AR support = { 8 } >
[     INFO ] < AR=y was requested for SL=8, but could not select SL with AR on ib_dev:1, SLs with AR support = { <none> }, SLs without AR support = { 8 } >
```

Reply:
IMO the part of message which tells how to fix is too long and may be confising. also should include mlx5_1-like dev name
can we make it more clear/simple?
e.g what does " Please, set SL manually without AR on ib_dev:1" mean?

Reply:
> e.g what does " Please, set SL manually without AR on ib_dev:1" mean?

it is from tests

Reply:
> IMO the part of message which tells how to fix is too long and may be confising. also should include mlx5_1-like dev name
> can we make it more clear/simple?

why it is confusing? it shows the name of the device, it shows the SL masks (for with AR and without AR), and what AR_ENABLE support was requested by user

Reply:
```
[     INFO ] < AR=n was requested for SL=auto, but could not detect AR mask for SLs. Please, set SL manually without AR on mlx5_0:1, SLs with AR support = { <none> }, SLs without AR support = { <none> } >
[     INFO ] < AR=n was requested for SL=8, but could not detect AR mask for SLs. Please, set SL manually without AR on mlx5_0:1, SLs with AR support = { <none> }, SLs without AR support = { <none> } >
[     INFO ] < AR=n was requested for SL=8, but could not select SL without AR on mlx5_0:1, SLs with AR support = { 8 }, SLs without AR support = { <none> } >
[     INFO ] < AR=n was requested for SL=8, but could not select SL without AR on mlx5_0:1, SLs with AR support = { 8 }, SLs without AR support = { <none> } >
[     INFO ] < AR=y was requested for SL=auto, but could not select SL with AR on mlx5_0:1, SLs with AR support = { <none> }, SLs without AR support = { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14 } >
[     INFO ] < AR=y was requested for SL=auto, but could not detect AR mask for SLs. Please, set SL manually with AR on mlx5_0:1, SLs with AR support = { <none> }, SLs without AR support = { <none> } >
[     INFO ] < AR=y was requested for SL=8, but could not select SL with AR on mlx5_0:1, SLs with AR support = { <none> }, SLs without AR support = { 8 } >
[     INFO ] < AR=y was requested for SL=8, but could not detect AR mask for SLs. Please, set SL manually with AR on mlx5_0:1, SLs with AR support = { <none> }, SLs without AR support = { <none> } >
[     INFO ] < AR=y was requested for SL=8, but could not select SL with AR on mlx5_0:1, SLs with AR support = { <none> }, SLs without AR support = { 8 } >
[     INFO ] < AR=y was requested for SL=8, but could not select SL with AR on mlx5_0:1, SLs with AR support = { <none> }, SLs without AR support = { 8 } >
```
changed `ib_dev:1` to `mlx5_0:1`